src/library/equations_compiler/elim_match.cpp

```diff
@@ -502,17 +503,14 @@ struct elim_match_fn {
                 expr const & p = head(eqn.m_patterns);
                 if (is_local(p)) {
                     has_variable = true; return true;
-                } else if (is_constructor_app(p)) {
+                }
+                type_context ctx = mk_type_context(eqn.m_lctx);
+                bool is_val = is_value(ctx, p);
+                if ((is_constructor_app(p) && (value_is_contructor || !is_val)) ||
+                    (value_is_contructor && is_val)) {
                     has_constructor = true; return true;
-                } else {
-                    type_context ctx = mk_type_context(eqn.m_lctx);
-                    if (is_value(ctx, p)) {
-                        if (value_is_contructor) has_constructor = true;
-                        return true;
-                    } else {
-                        return false;
-                    }
                 }
+                return false;
             });
         return r && has_variable && has_constructor;
     }
```

Comment:
There is some fallout from this change in form of 'enum' inductives being compiled with `ite` and then failing in the app builder because they don't have a `decidable_eq` instance. What should we do about that?